Feature/extract strings cx 1551

stefaniacardenas opened the pull request:
This PR sets the infrastructure for the internationalisation of the JS SDK that will allow for multiple language use.
I have extracted all of our strings into `locales/en.js` and they get injected into the code using `polyglot.js`

rfreitas commented:
Travis automatic deployment: https://staging-285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://release-285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://staging-285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://release-285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://release-285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://staging-285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://285-pr-onfido-sdk-ui-onfido.surge.sh

hjheath commented on README.md:
````diff
@@ -214,6 +214,17 @@ A number of options are available to allow you to customise the SDK:
+    ```javascript
+    language: {
+      locale: 'es', // required
+      customStrings: {some_key: 'This is my custom string'}
````
I don't think we should allow runtime customisation of strings. It adds complexity and I don't imagine anyone will actually use it.

@rfreitas @seewah thoughts?

rfreitas replied:
I agree, I would remove it for now.

hjheath commented on README.md:
````diff
@@ -214,6 +214,17 @@ A number of options are available to allow you to customise the SDK:
+
+    ```javascript
+    language: {
+      locale: 'es', // required
````
This isn't required though right? We default to English

stefaniacardenas replied:
Correct. That's not required. It's an example of how the language object should look if they want to specify a locale that is not English.

rfreitas replied:
Interesting that you already done this, but I'm not sure if the feature should already be there (or maybe advertised) considering that no other language can be chosen.

stefaniacardenas replied:
Yes, you are right. I was thinking that this PR and the Spanish translation one would be merged and released together but this PR can actually be merged independently, so I could just remove the changes in the `README.md` and `CHANGELOG.md` and include them in the next PR

hjheath commented on locales/index.js:
```diff
@@ -0,0 +1,9 @@
+
+const locales = {
+  en,
+  es
```
What's the schema for these? I think it should be a [language tag](https://www.w3.org/International/questions/qa-lang-2or3), but we should make that clear.

stefaniacardenas replied:
Yes, I can add a comment.

rfreitas replied:
does es exist yet?

stefaniacardenas replied:
It's just an empty object for now. I've used it to try `polyglot.js` features at the beginning and I included in this PR as I thought that the Spanish translation PR would be merged into this branch and merge all the code at once.

hjheath commented on src/index.js:
```diff
@@ -61,11 +64,20 @@ const formatOptions = ({steps, ...otherOptions}) => ({
   steps: (steps || ['welcome','document','face','complete']).map(formatStep)
 })
 
+const setI18n = (options) => {
```
Nice - we estimated this as part of another issue, but you've already done it!

rfreitas commented:
Travis automatic deployment: https://release-285-pr-onfido-sdk-ui-onfido.surge.sh

hjheath commented on src/index.js:
```diff
@@ -61,11 +64,20 @@ const formatOptions = ({steps, ...otherOptions}) => ({
+  const locale = options.language ? options.language.locale : 'en'
+  const customStrings = options.language ? options.language.customStrings : null
+  const phrases = locales[locale]
+  const polyglot = new Polyglot({locale, phrases})
```
why do we need to pass the locale in?

stefaniacardenas replied:
That's only required to handle pluralisation, but we aren't pluralising any string at the moment. I can remove it for now and re-introduce it if we need it for Spanish translations.

rfreitas commented:
Travis automatic deployment: https://staging-285-pr-onfido-sdk-ui-onfido.surge.sh

hjheath commented on src/index.js:
```diff
@@ -61,11 +64,20 @@ const formatOptions = ({steps, ...otherOptions}) => ({
   Tracker.track()
-  const options = formatOptions({ ...defaults, ...opts, events })
+  const i18n = setI18n(opts)
+  const options = formatOptions({ ...defaults, ...opts, events, i18n })
```
Do we need to pass it down to all components? Would it work as a global utility function?

stefaniacardenas replied:
Yes, I could use a function to utility function to initialise polyglot and a HOC to wrap up all the components with the translations.

stefaniacardenas replied:
I was reading [this article](https://www.jonathan-petitcolas.com/2017/03/27/internationalizing-react-application-using-polyglot.html) that tackles this probles using HOC and context. It seems like a good way to fix this issue. I would do it as part of another story though.

hjheath replied:
Interesting article. What if we just use a global?

stefaniacardenas replied:
Uhm, I am a bit wary of using a global function, seems like code smell to me because 2 globals with the same name might break the code and it seems safer to use this solution. What are your thoughts @rfreitas?

stefaniacardenas replied:
Unless you mean that this code could be extracted into a utils function, but that would mean that we would have to instantiate `polyglot` every time we import it.

hjheath replied:
Could we have a global `i18n` object in utils? Then we can just import it and call `i18n.t` whenever we need a string?

I'm not sure how you would instantiate that object without using the window, but it would make passing `i18n` around all over the place go away.

Regardless I think we should do some thinking to avoid passing it around. I think we should do it as part of this PR because adding `i18n` to all the props is probably unnecessary.

rfreitas commented on src/components/Camera/index.js:
```diff
@@ -99,14 +99,15 @@ export default class Camera extends Component {
-      onFallbackClick: () => this.stopCamera}
-    }
+      onFallbackClick: () => this.stopCamera,
+      desktopCopy: uploadDesktop(i18n)
```
what is `uploadDesktop`?

rfreitas commented on locales/en.js:
```diff
@@ -0,0 +1,187 @@
+    title: 'Verify your identity',
+    hint: 'Select the type of document you would like to upload',
+  },
+  upload_desktop: {
```
there seem to be redundant strings. for sure something we need to avoid, otherwise with more languages it will become very hard to manage. Denesting might be preferable as well, since that way we can decouple logic from strings as much as possible.

rfreitas replied:
there are some techniques that Android uses for this scenario (desktop vs mobile) that we could mimic

rfreitas commented on src/index.js:
```diff
@@ -61,11 +64,20 @@ const formatOptions = ({steps, ...otherOptions}) => ({
   steps: (steps || ['welcome','document','face','complete']).map(formatStep)
 })
 
+const setI18n = (options) => {
```
any reason for calling it `i18n` if we are not using `i18n`? is it a standard term on js?

stefaniacardenas replied:
`i18n` stands for `internationalization` and this function is just setting the internationalization. Therefore I've decided to use `i18n` independently of the library we use as it's more generic and representative of what the function does.

rfreitas commented on src/components/strings/index.js:
```diff
@@ -1,104 +1,103 @@
-export const uploadDesktop = {
+export const uploadDesktop = (i18n) => ({
```
why can't use just the local file straight away?
A lot of redundancy redeclaring everything again.

stefaniacardenas replied:
That's what I did in my first implementation but the keys were becoming heavily interpolated and I wasn't sure it was a good idea to delegate conditional logic to the presence of a key/copy, which made decided for this solution. At a closer look, I think having this redundancy may generate bugs in the long run, therefore I will find a better way to handle the interaction between copy and logic.

rfreitas commented:
I think you deleted the `package.lock` by mistake

hjheath commented:
Did you mean to remove `package-lock.json`?

rfreitas commented on locales/en.js:
```diff
@@ -0,0 +1,180 @@
+  upload_mobile: {
+    driving_licence: {
+      front: {
+        title: 'Front of license',
```
redundant string

stefaniacardenas replied:
Yes, I'm actually in the process of drying up the whole file by using interpolation

rfreitas commented:
Travis automatic deployment: https://staging-285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://release-285-pr-onfido-sdk-ui-onfido.surge.sh

stefaniacardenas commented:
Latest changes:
* removed interpolation
* separate default locales and mobile locales
* mobile locales will override default strings if is not desktop
* removed references to `es`
* removed string override support
* added package-lock.json

rfreitas commented:
Travis automatic deployment: https://285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://release-285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://staging-285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://staging-285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://release-285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented on src/index.ejs:
```diff
@@ -85,7 +85,11 @@
           },
-          steps: steps
+          steps: steps,
+          language: {
```
is this object required? it shouldn't be.

stefaniacardenas replied:
No, it's not required

rfreitas commented on locales/mobileLocales/en.js:
```diff
@@ -0,0 +1,33 @@
+    driving_licence: {
+      front: {
+        instructions: 'Take a photo of the front of your license',
+        webcam: ''
```
Minor: do we need it? the component shouldn't ask for strings if he doesn't want to display them

rfreitas commented on locales/mobileLocales/en.js:
```diff
@@ -0,0 +1,33 @@
+export const en = {
```
nicely done!

rfreitas commented on src/index.js:
```diff
@@ -61,11 +65,19 @@ const formatOptions = ({steps, ...otherOptions}) => ({
+  const locale = options.language ? options.language.locale : 'en'
+  const phrases = locales[locale]
+  const polyglot = new Polyglot({locale, phrases, onMissingKey: () => null})
+  if (!isDesktop) polyglot.extend(mobileLocales[locale])
```
👍

rfreitas commented:
Travis automatic deployment: https://staging-285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://release-285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://285-pr-onfido-sdk-ui-onfido.surge.sh

stefaniacardenas commented:
Updates:
* removed `language` from options
* removed empty string from mobile locales, handling logic within component

stefaniacardenas commented:
I've created a [Jira ticket](https://jira.onfido.co.uk/browse/CX-1621) to investigate the best way to make the polyglot instance globally available.

rfreitas commented:
Travis automatic deployment: https://staging-285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://release-285-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://285-pr-onfido-sdk-ui-onfido.surge.sh
